nixos/modules/services/networking/amuled.nix

```
@@ -1,89 +1,506 @@
+      IncomingDir = mkOption {
+        type = types.path;
+        default = "${cfg.dataDir}/Incoming";
+        defaultText = literalExpression "\${config.services.amule.dataDir}/Incoming";
         description = ''
-          The user the AMule daemon should run as.
+          Directory where aMule moves completed downloads.
+          Files in this directory are automatically shared.
+          Ensure the aMule service has write permissions
         '';
       };
+      OSDirectory = mkOption {
+        type = types.path;
+        default = cfg.dataDir;
+        defaultText = literalExpression "\${config.services.amule.dataDir}";
+        description = "On-disk state directory, probably you don't want to change this";
+      };
+    };
+    ExternalConnect = {
+      ECPort = mkOption {
+        type = types.port;
+        default = 4712;
+        description = "TCP port for external connections, like remote control via amule-gui";
+      };
+      ECPassword = mkOption {
+        type = types.str;
+        default = "";
+        description = ''
+          MD5 hash of the password, obtainaible with `echo "<password>" | md5sum | cut -d ' ' -f 1`
+        '';
+      };
+    };
+    WebServer = {
+      Enabled = lib.mkOption {
+        type = types.enum [
+          0
+          1
+        ];
+        default = 0;
+        description = "Set to 1 to enable the web server";
+      };
+      Password = mkOption {
+        type = types.str;
+        default = "";
+        description = ''
+          MD5 hash of the password, obtainaible with `echo "<password>" | md5sum | cut -d ' ' -f 1`
+        '';
+      };
+      Port = mkOption {
+        type = types.port;
+        default = 4711;
+        description = "Web server port";
+      };
+    };
+  };
+
+  finalSettings = recursiveUpdate defaultSettings cfg.settings;
+
+  webServerEnabled = cfg.settings.WebServer.Enabled == 1;
+in
+{
+  options.services.amule = {
+    enable = mkEnableOption "aMule daemon";
 
+    package = mkPackageOption pkgs "amule-daemon" { };
+
+    amuleWebPackage = mkPackageOption pkgs "amule-web" { };
+
+    extraArgs = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      description = "Additional passed arguments";
+    };
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/amuled";
+      description = "Directory holding configuration and by default also incoming and temporary files";
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "amule";
+      description = "The user the aMule daemon should run as";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "amule";
+      description = "Group under which amule runs";
+    };
+
+    openPeerPorts = mkEnableOption "open the peer port(s) in the firewall";
+
+    openExternalConnectPort = mkEnableOption "open the external connect port";
+
+    openWebServerPort = mkEnableOption "open the web server port";
```

```suggestion
    openFirewall = mkEnableOption "opening the aMule ports in the firewall";
```
Any reason why this can't be combined into one `openFirewall` setting? Enabling it will open the aMule ports, and if the webserver is also enabled, then it'll open the webserver port as well.